graphs/greedy_best_first.py

```diff
@@ -0,0 +1,162 @@
+
+        self.open_nodes = [self.start]
+        self.closed_nodes = []
+
+        self.reached = False
+
+        self.path = [(self.start.pos_y, self.start.pos_x)]
+        self.costs = [0]
+
+    def search(self):
+        while self.open_nodes:
+            # Open Nodes are sorted using __lt__
+            self.open_nodes.sort()
+            current_node = self.open_nodes.pop(0)
+
+            if current_node.pos == self.target.pos:
+                self.reached = True
+                self.path = self.retrace_path(current_node)
+                break
+
+            self.closed_nodes.append(current_node)
+            successors = self.get_successors(current_node)
+
+            for child_node in successors:
+                if child_node in self.closed_nodes:
+                    continue
+
+                if child_node not in self.open_nodes:
+                    self.open_nodes.append(child_node)
+                else:
+                    # retrieve the best current path
+                    better_node = self.open_nodes.pop(self.open_nodes.index(child_node))
+
+                    if child_node.g_cost < better_node.g_cost:
+                        self.open_nodes.append(child_node)
+                    else:
+                        self.open_nodes.append(better_node)
+
+        if not (self.reached):
+            print("No path found")
+
+    def get_successors(self, parent: Node) -> List[Node]:
+        """
+        Returns a list of successors (both in the grid and free spaces)
+        """
+        successors = []
+        for action in delta:
+            pos_x = parent.pos_x + action[1]
+            pos_y = parent.pos_y + action[0]
+            if (
+                pos_x < 0
+                or pos_x > len(grid[0]) - 1
+                or pos_y < 0
+                or pos_y > len(grid) - 1
+            ):
+                continue
+
+            if grid[pos_y][pos_x] != 0:
+                continue
+
+            node_ = Node(
+                pos_x,
+                pos_y,
+                self.target.pos_y,
+                self.target.pos_x,
+                parent.g_cost + 1,
+                parent,
+            )
+            successors.append(node_)
+        return successors
+
+    def retrace_path(self, node: Node) -> List[Tuple[int]]:
+        """
+        Retrace the path from parents to parents until start node
+        """
+        current_node = node
+        path = []
+        while current_node is not None:
+            path.append((current_node.pos_y, current_node.pos_x))
+            current_node = current_node.parent
+        path.reverse()
+        return path
+
+
+# all coordinates are given in format [y,x]
+init = (0, 0)
+goal = (len(grid) - 1, len(grid[0]) - 1)
+for elem in grid:
+    print(elem)
+
+print("------")
+
+greedy_bf = GreedyBestFirst(init, goal)
+greedy_bf.search()
+
+for elem in greedy_bf.path:
+    grid[elem[0]][elem[1]] = 2
+
+for elem in grid:
+    print(elem)
```

Comment:
Python files should end with one and only one `\n`.  Doing this will get rid of that GitHub 🚫